crates/core/src/checkpoint.rs

```diff
@@ -106,59 +125,105 @@ impl CheckPoint {
         };
 
         CheckPoint::new(prev_block_id)
-            .push(this_block_id)
+            .push_block_id(this_block_id)
             .expect("must construct checkpoint")
     }
 
-    /// Puts another checkpoint onto the linked list representing the blockchain.
+    /// Construct a checkpoint from a list of [`BlockId`]s in ascending height order.
     ///
-    /// Returns an `Err(self)` if the block you are pushing on is not at a greater height that the one you
-    /// are pushing on to.
-    pub fn push(self, block: BlockId) -> Result<Self, Self> {
-        if self.height() < block.height {
-            Ok(Self(Arc::new(CPInner {
-                block,
-                prev: Some(self.0),
-            })))
-        } else {
-            Err(self)
+    /// # Errors
+    ///
+    /// This method will error if any of the follow occurs:
+    ///
+    /// - The `blocks` iterator is empty, in which case, the error will be `None`.
+    /// - The `blocks` iterator is not in ascending height order.
+    /// - The `blocks` iterator contains multiple [`BlockId`]s of the same height.
+    ///
+    /// The error type is the last successful checkpoint constructed (if any).
+    pub fn from_block_ids(
+        block_ids: impl IntoIterator<Item = BlockId>,
+    ) -> Result<Self, Option<Self>> {
```

Comment:
Let's rework this method to be generic over `D`:

```suggestion
    pub fn from_blocks(
        blocks: impl IntoIterator<Item = (u32, D)>,
    ) -> Result<Self, Option<Self>> {
```